[푸코] Step02_3 - 리팩토링

wnsxor1993 opened the pull request:
## 결과 화면
![화면 기록 2022-03-03 오후 3 58 53](https://user-images.githubusercontent.com/44107696/156513460-4e134ce6-623b-4bd8-9ec9-d8b2561640e0.gif)

## 작업 목록
- [x] Step01 피드백 반영
    - [x] 타입 속성 파일 분할
    - [x] RGB 범위 추가
    - [x] 타입 자체 전달 형태로 메서드 수정
    - [x] RectangleFactory / RectangleIdFactory 분할
 - [x] RandomRectange 생성 기능 담당하는 RandomRectangleMaker 구현
 - [x] Plane 구조체 기능 구현
 - [x] 우측 설정 뷰 / 사각형 뷰 커스텀 뷰로 구현
 - [x] GestureRecognizer를 통해 사각형 속성 표출 
 - [x] addTarget을 활용해 Slider 값 변경과 모델의 데이터 연결 
     - [x] custom delegate를 통한 VC와의 연결 구현
 - [x] 전체 로직 리팩토링
     - [x] 팩토리 간소화
     - [x] 뷰 / 모델 관계성 및 역할 집중   

## 학습 키워드
1. 커스텀 UIView
2. UIGestureRecognizerDelegate (hitTest)
3. UISlider
4. addTarget
5. custom delegate
6. withAlphaComponent

## 고민과 해결
- 처음에는 커스텀 뷰에서 액션을 연결하고 이를 VC에 전달하는 방법이 어려워 VC에서 직접 해당 뷰의 슬라이더에 접근하여 액션을 연결해주는 방식을 사용했습니다. 하지만 하위 뷰로 선언하였으니 액션을 타겟하는 것도 해당 커스텀 뷰 내부에서 구현하고 이를 한 번에 VC에 넘기는 방식을 지향해야 된다는 얘기를 듣고 다시금 방법을 찾게 되었고 커스텀으로 delegate를 구현하는 방식을 배우게 되었습니다.
이에 해당 커스텀 뷰의 액션 타겟과 VC를 커스텀 delegate를 통해 연결시키는 방식으로 수정할 수 있었습니다. 다만 이 과정에서 커스텀 뷰의 레이아웃을 생성하는 함수 호출을 지워버렸고 이로인해 생긴 문제를 찾아서 해결하는데에 또 많은 시간을 보내게 되었습니다.
덕분에 수정할 때, 좀 더 세밀하게 코드를 건드려야된다는 것을 마음 깊게 배울 수 있었습니다.

--------------------------------------------------------------------------------------------
- 우측의 속성 표출 관련 뷰를 스토리보드에서 만든 뒤, IB로 연결하는 방식을 하려고 했지만 커스텀 뷰에도 어느정도 익숙해지는 게 필요할 듯 싶어 기존에 작성한 사각형 뷰와 함께 커스텀 뷰 형태로 구현해보았습니다.

- 객체가 구현하는 기능을 비대하게 만드는 것보다 하나에 집중할 수 있도록 구현하는 것이 좋을 듯 했고, 이전 리뷰에서도 IDFactory에 대한 언급이 있어 Factory를 ID와 사각형 두 가지로 나누되 이를 종합해서 완성된 형태의 랜덤 사각형을 만들어내는 RandomRectangleMaker를 추가하였습니다.

- 커스텀 뷰 내의 Slider의 value 변경과 모델의 데이터 연동을 하기 위해 처음에는 커스텀 뷰에 Rectangle 인스턴스를 넘겨주려 했지만 그렇게 되면 View가 모델을 소유하고 데이터에 대한 변경도 담당하는 형태가 되어버리기에 다른 방법을 다시 찾아봤습니다.
그러던 중, 커스텀 뷰 또한 VC가 MainViewController인 것이 생각났고 @IBOutlet으로 내부 view들을 연결하는 방식이나 직접 프로퍼티로 선언된 view를 호출하는 방식에 큰 차이가 없다고 생각이 들어 MainViewController에서 커스텀 뷰 내에 선언된 Slider 뷰를 직접 호출하여 addTarget을 구현했고, 해당 메서드의 action에 인자로 넘긴 @objc 함수에 모델의 데이터 변경 로직을 기입하여 VC가 모델의 데이터를 변경하는 형태가 되도록 구현하였습니다.

- Slider에서 value를 받아오면 해당 값은 Float 타입이라 Alpha 타입과의 매칭 시, Double로의 타입 변환이 필요하고 이로 인해 값에 미묘한 변화가 생겨 올바르게 매칭되지 못해 사각형의 alpha 속성 변화가 제대로 이뤄지지 않았습니다.
여러 방법을 시도해도 변화가 없었고, Double 타입으로 변환된 value가 0.100000392 같은 형태로 변질되는 점을 보고 switch를 활용하여 범위 내의 값만 맞으면 올바른 Alpha 타입을 리턴할 수 있도록 changeAlpha 함수의 로직을 구현하였습니다.

wnsxor1993 committed:
Chore: 일부 구현 요소 수정
- RGB init 범위 예외 요소 구현
- Factory 기능 분할
- 원시 타입 직접 전달 최소화

wnsxor1993 committed:
Feat: 좌표 판단 기능 추가
- 기존 RectangleArray 클래스를 요구사항에 맞춰 Plane 구조체로 변경
- Plane 구조체에 좌표를 통해 사각형 존재 여부 기능 추가

wnsxor1993 committed:
Feat: 랜덤 사각형 생성 객체 구현
- Plane은 생성된 사각형들을 가지고 관리하는 기능만으로 한정
- 랜덤한 사각형을 생성하고 전달하는 기능은 RandomRectangleMaker 객체에게 부여

wnsxor1993 committed:
Feat: Custom View 구현
- 랜덤 사각형, 우측 설정 관련 View를 커스텀 클래스로 구현

wnsxor1993 committed:
Feat: 사각형 터치 시 데이터 표출 기능 구현
- GestureRecognizer를 통해서 사각형 터치 시 우측 view에서 정보 표출 구현

wnsxor1993 committed:
Feat: UnitTest 추가
- 신규 작성 Model 로직 관련 UnitTest 추가 및 확인

godrm commented:
PR에서는 당연히 시도하지 않은걸, 찾지못한걸 물어본다고 해서 알려주지는 않습니다. 
작성한 코드에 대한 의견을 나누는 공간이고 시도한 과정에 대한 의견을 주고 받는 곳입니다. 
하지 않은 방법에 대해 수업하는 역할이 아닙니다. 참고하세요~

wnsxor1993 committed:
Feat: 속성 변경 기능 구현
- 선택된 사각형의 RGB와 Alpha 값을 slider를 통해 변경 가능

wnsxor1993 commented:
넵! 말씀하신 내용 명심하겠습니다^^
속성 변경 기능 추가하고 이에 맞춰 PR 수정하였습니다.

godrm commented on DrawingApp/Model/Attributes/MyPoint.swift:
```diff
@@ -0,0 +1,26 @@
+    }
+    
+    func yValue() -> Double{
+        return self.y
```
struct - let으로 선언한 경우는 이렇게 getter를 제공할 필요가 전혀 없습니다

godrm commented on DrawingApp/Model/Attributes/RGBColor.swift:
```diff
@@ -0,0 +1,47 @@
+            self.blue = 255
+        } else{
+            self.blue = blue
+        }
```
타입을 생성할 때 어떤 식으로 구현할지를 선택해야 하는데,  
• 범위를 벗어난 값이 들어와도 범위로 잘라서 만들어 줄 것인가  
• 혹은 아예 만들어지지 않도록 예외처리를 할 것인가  
• 아예 넣지도 못하도록 할 것인가  
여러가지 방법이 있습니다. 그 중에서도 개발자들이 실수를 줄일 수 있고, 런타임에 실행하지 않고 컴파일 타임에 점검할 수 있으면 가장 좋습니다. 참고하세요

godrm commented on DrawingApp/Model/Rectangle/Plane.swift:
```diff
@@ -0,0 +1,56 @@
+        var findedRectangle: Rectangle?
+        
+        rectangles.forEach{ rectangle in
+            let minX: Double = rectangle.showPoint().xValue()
+            let maxX: Double = minX + rectangle.showSize().widthValue()
+            let minY: Double = rectangle.showPoint().yValue()
+            let maxY: Double = minY + rectangle.showSize().heightValue()
```
struct로 값만 제공하는 타입의 경우는 속성을 쓰셔도 됩니다. 이게 더 복잡한 코드가 되니까요 
다만 계속 얘기하는 것처럼 하위 rectangle 속성을 가져와서 여기서 다 비교하려고 하기 보다는,
rectangle에 필요한 값을 넘겨서 판단해주는 메소드를 만들려고 하세요

godrm commented on DrawingApp/Model/Rectangle/RandomRectangleMaker.swift:
```diff
@@ -0,0 +1,25 @@
+    func makeRectangle(viewWidth: Double, viewHeight: Double) -> Rectangle{
+        let id = rectangleIDFactory.makeID()
+        let size = rectangleFactory.makeSize()
+        let point = rectangleFactory.makePoint(viewWidth: viewWidth, viewHeight: viewHeight)
```
`makePoint(viewWidth:, viewHeight:)`는 어떤 의미인가요? viewWidth와 Height 크기를 갖는 Point를 생성하는 걸까요?

wnsxor1993 replied:
다른 뷰들이 배치되어 있지 않은 부모 뷰의 너비와 길이 값을 받아와 해당 범위 내에서만 생성되도록 만들기 위한 매개변수입니다. 의미가 제대로 전달되고 있지 않는 듯 한데, 매개변수명을 다시 고민해보도록 하겠습니다.

godrm commented on DrawingApp/Model/Rectangle/RandomRectangleMaker.swift:
```diff
@@ -0,0 +1,25 @@
+
+class RandomRectangleMaker{
+    private let rectangleFactory = RectangleFactory()
+    private let rectangleIDFactory = RectangleIDFactory()
```
보통 팩토리는 인스턴스를 생성하지 않고 가볍게 필요한 것만 만드는 경우가 많습니다.
인스턴스를 생성하는 순간이 관리할 부분이 늘어나게 되니까요

godrm commented on DrawingApp/Model/Rectangle/Rectangle.swift:
```diff
@@ -0,0 +1,78 @@
+        return description
+    }
+    
+    func showId() -> String{
+        return id
+    }
+    
+    func showSize() -> MySize{
+        return size
+    }
+    
+    func showPoint() -> MyPoint{
+        return point
+    }
+    
+    func showColor() -> RGBColor{
+        return color
+    }
+    
+    func showAlpha() -> Alpha{
+        return alpha
+    }
+    
+    func changeColor(color: RGBColor){
+        self.color = color
+    }
```
private var로 지정한다고 해서 모든 값을 다 getter를 만드는 것은 지양하세요. 
이거 계속 얘기하는 것 같은데 속성을 가져가지 않도록 하는게 중요합니다.

wnsxor1993 replied:
해당 부분에 대한 의미가 아직 명확하게 이해되지 않습니다.
속성을 가져가지 않는 것이 Int나 Double과 같은 원시 타입의 값을 직접적으로 넘겨주지 않고 캡슐화된 데이터 형태로 넘겨주는 것으로 이해해도 괜찮을까요?
추가적으로 원시 타입의 값은 최하위의 데이터 구조에서만 직접 소유하는 형태를 지향해야 하는 걸까요?

godrm replied:
캡슐화하는 타입과 아닌 타입이 있는데, 값 타입으로 만든 경우는 그냥 struct - let 구조로 두면 됩니다.
그런 타입들까지 캡슐화를 하라는 건 아니에요 
원시 타입들을 직접 사용하는 경우는 줄이는 방향이 맞구요, 그걸 객체로 표현하느냐 값 타입으로 표현하느냐 다양한 접근이 있는 겁니다.

godrm commented on DrawingApp/Model/Rectangle/Rectangle.swift:
```diff
@@ -0,0 +1,78 @@
+        case 0.0..<0.2:
+            self.alpha = .one
+        default:
+            self.alpha = .ten
```
이 함수는 왜 Rectangle이 이걸 구현해야 하는지 잘 모르겠습니다. 
매개변수가 Double이 되는 순간 작업해야 하는 내용이 결정되어 버리는 것 같네요
Alpha 타입을 선언했으면 그냥 Alpha 타입을 쓰도록 하는게 좋을 것 같습니다.

godrm commented on DrawingApp/Model/Rectangle/RectangleFactory.swift:
```diff
@@ -7,15 +7,10 @@
-    
     func makeSize() -> MySize{
         let size = MySize(width: self.width, height: self.height)
         return size
```
RectangleFactory와 RandomRectangleMaker 차이는 뭔지 궁금합니다.

wnsxor1993 replied:
Rectangle 속성과 Rectangle Id를 리턴해주는 기능을 가진 팩토리를 소유하여 랜덤한 Rectangle을 생성해주는 용도로 작성하였습니다. 다만 앞에서 말씀하신 팩토리의 경량화를 고려하니 해당 클래스를 굳이 유지할 필요가 없어보여 VC의 버튼 액션에서 Rectangle 생성을 구현하는 방향으로 수정하려 합니다.

godrm commented on DrawingApp/Model/Rectangle/RectangleIDFactory.swift:
```diff
@@ -0,0 +1,29 @@
+
+final class RectangleIDFactory{
+    func makeID() -> String{
+        let alphabet: [String] = ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z", "1", "2", "3", "4", "5", "6", "7", "8", "9"]
```
이렇게 값을 하드코딩하기 보다는 이미 있는 타입이 있지 않나 찾아보세요
CharacterSet 이란 타입도 도움이 될 것 같습니다

godrm commented on DrawingApp/Model/Rectangle/RectangleIDFactory.swift:
```diff
@@ -0,0 +1,29 @@
+final class RectangleIDFactory{
+    func makeID() -> String{
+        let alphabet: [String] = ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z", "1", "2", "3", "4", "5", "6", "7", "8", "9"]
+        var name: [String] = []
```
문자열 배열 타입인데 name 단수로 표현하다보니 헷갈리는 것 같습니다

godrm commented on DrawingApp/Model/Rectangle/RectangleIDFactory.swift:
```diff
@@ -0,0 +1,29 @@
+        let alphabet: [String] = ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "r", "s", "t", "u", "v", "w", "x", "y", "z", "1", "2", "3", "4", "5", "6", "7", "8", "9"]
+        var name: [String] = []
+        
+        for num in 0..<11{
```
num 같은 단어도 역할을 좀 더 구체적으로 표현하는 단어를 그대로 써보세요.

wnsxor1993 commented:
> PR에서는 당연히 시도하지 않은걸, 찾지못한걸 물어본다고 해서 알려주지는 않습니다. 작성한 코드에 대한 의견을 나누는 공간이고 시도한 과정에 대한 의견을 주고 받는 곳입니다. 하지 않은 방법에 대해 수업하는 역할이 아닙니다. 참고하세요~

godrm commented on DrawingApp/VC/MainViewController.swift:
```diff
@@ -0,0 +1,102 @@
+        rightAttributerView.redSlider.addTarget(self, action: #selector(self.changeRectangleAttribute), for: .valueChanged)
+        rightAttributerView.greenSlider.addTarget(self, action: #selector(self.changeRectangleAttribute), for: .valueChanged)
+        rightAttributerView.blueSlider.addTarget(self, action: #selector(self.changeRectangleAttribute), for: .valueChanged)
+        rightAttributerView.alphaSlider.addTarget(self, action: #selector(self.changeRectangleAttribute), for: .valueChanged)
```
RightAttributerView를 하위 뷰로 선언했으면 이런 타겟들도 RightAttributerView 내부에서 처리하고,
한꺼번에 알려줄 수 있는 방법을 고려해보세요

godrm commented on DrawingApp/VC/MainViewController.swift:
```diff
@@ -0,0 +1,102 @@
+        rectangleView.makeRectangleView(width: self.rightAttributerView.frame.minX, height: self.rectangleButton.frame.minY)
+        
+        guard let rectangle = rectangleView.giveRectangle() else{
+            let alert = UIAlertController(title: "Warning", message: "사각형 생성에 문제가 있습니다.", preferredStyle: .actionSheet)
+            let action = UIAlertAction(title: "Okay", style: .cancel)
+            alert.addAction(action)
+            self.present(alert, animated: false)
```
이런 경우는 뭔지 모르겠지만, 이게 사용자가 잘못한 걸까요? giveRectangle() 이게 의미하는 게 무엇인지 모르겠습니다. 
Alert가 필요한 경우가 있지만, 사용자 UX를 막아서기 때문에 조심해서 써야 하는 UI입니다.

godrm commented on DrawingApp/VC/MainViewController.swift:
```diff
@@ -0,0 +1,102 @@
+    }
+    
+    @objc func changeRectangleAttribute(){
+        guard let index = selectedRectangleIndex, let rectangle = self.rectangles[index], let rectView = selectedRectangleView else{
+            return
+        }
+        let newColor = RGBColor(red: rightAttributerView.redValue, green: rightAttributerView.greenValue, blue: rightAttributerView.blueValue)
+        let newAlpha = rightAttributerView.alphaValue
+        
+        rectangle.changeAlpha(alpha: newAlpha)
+        rectangle.changeColor(color: newColor)
+        
+        rectView.backgroundColor = UIColor(red: rectangle.showColor().redValue(), green: rectangle.showColor().greenValue(), blue: rectangle.showColor().blueValue(), alpha: rectangle.showAlpha().showValue())
+        
+        rightAttributerView.changeAllSliderValue()
```
속성을 바꾸면 색상과 알파값을 각각 처리하지 않고 모두다 바꾸는 건가요? 어느 하나만 바뀌도록 분리하는 것과 합치는 게 어떤 차이가 있을까요?

wnsxor1993 replied:
색상을 변경할 때, 알파 값이 항상 들어가다보니 따로 분리할 생각 없이 한 번에 처리하는 형태로 구현했습니다. 다만 해당 경우에는 값이 수정될 때마다 값 복사나 인스턴스 생성 등에 있어 지속적으로 불필요하게 이루어집니다.
이에 관련된 사항들을 알파와 색상의 변경에 따라 각각의 작업을 요청하는 형태의 로직으로 수정하였습니다.

godrm commented on DrawingApp/VC/MainViewController.swift:
```diff
@@ -0,0 +1,102 @@
+            return
+        }
+        
+        self.view.subviews.forEach{ rectangle in
+            guard rectangle.restorationIdentifier == rectangleInPlane.showId() else{
+                return
+            }
+            self.selectedRectangleView = rectangle
+            self.selectedRectangleIndex = rectangles.findRectangleIndex(rectangle: rectangleInPlane)
+        }
```
이런 뷰를 찾는 코드도 UIView 요소에서 찾는 것은 좋은 방법이 아닙니다. Plane에서 직접 찾을 수 있는게 가장 좋습니다.

wnsxor1993 replied:
모델 데이터의 값을 변경하거나 서칭 및 계산하는 등의 로직은 VC 내에 직접 구현하는 것이 아닌, 모델 내에 구현된 메서드들을 통해 진행하는 부분에 대해서는 알겠습니다만 UIView를 찾는 로직을 VC가 아닌 모델에서 진행하는 것이 좋은 이유를 아직 잘 이해되지 않습니다.
VC는 모델 데이터를 직접 소유하며 액션에 따라 모델과 뷰의 변화를 전달하는 역할 외에는 사소한 기능도 직접 구현하는 것을 지양해야 되고, View는 액션에 따른 변화 요소를 받아와 자신을 그려내는 데에만 집중해야 하므로 이러한 로직은 상위 모델에서 구현되는 것을 지향해야 하는 걸까요?

godrm replied:
아.. 오해가 있는 것 같네요. `UIView 목록에서 이벤트를 처리할 모델을 찾지말라`는 것이고 대신 이벤트를 처리할 모델을 찾는 것은 Plane에서 하라는 겁니다. 어떤 뷰가 선택됐느냐를 찾는 것도 결국 그 좌표를 포함하는 모델이 뭘까?를 찾는거라고 생각하시면 됩니다.

godrm commented on DrawingApp/View/RandomRectangleView.swift:
```diff
@@ -0,0 +1,50 @@
+        self.frame = CGRect(x: rectangle.showPoint().xValue(), y: rectangle.showPoint().yValue(), width: rectangle.showSize().widthValue(), height: rectangle.showSize().heightValue())
+    }
+    
+    func makeRectangleView(width: Double, height: Double){
+        self.rectangle = randomRectangle.makeRectangle(viewWidth: width, viewHeight: height)
+        
+        layout()
+        attribute()
```
이 메소드는 누가 언제 불러주는 걸까요? init()이 생성자인데 생성자와 전혀 상관없이 생성하는 것 같네요
그리고 View가 Rectangle을 소유하면 안됩니다. 
가볍게 만들라는 의미는 모델을 직접 가지지 말고, 그리는 역할에만 집중하세요 
팩토리 타입을 참조해도 안됩니다. 뷰는 뷰 역할에만 집중해보세요

wnsxor1993 replied:
생성 위치로 인해 상위 뷰의 비어있는 공간 범위를 알아야 속성이나 레이아웃의 설정이 가능하다보니 MainViewController에서 해당 메서드를 호출하는 방식으로 구현하였습니다.
다만 View가 모델을 소유하면 안 된다는 점에 맞춰 수정을 하다보니 attribute와 layout에 Rectangle 인스턴스를 인자 값으로 넘겨주는 형태로 바꾼 뒤, VC에서 RandomRectangleView의 인스턴스를 생성해서 직접 두 메서드를 호출하여 속성과 레이아웃을 설정하는 형태로 수정했고 상기의 메서드는 제거하였습니다

godrm reviewed (changes requested):
View와 Model의 관계에 대해 피드백을 확인해보세요. 
뷰는 뷰 역할에만 집중해야 합니다.

wnsxor1993 committed:
Chore: Model, View 역할 분배 수정
- getter 최소화, 상하위 모델 기능 세분화
- View 역할에 맞춰 기능 구현 재분배

wnsxor1993 committed:
Chore: ID 생성 기능 수정
- UUID를 활용하여 ID를 생성하는 로직으로 수정

wnsxor1993 committed:
Feat: Delegate 추가
- RightAttributerView Delegate 추가하여 Slider 액션 구현
- VC 내에 구현한 UIView 서칭 기능 Plane 내로 변경

godrm commented:
위에 코멘트들이 모두 반영이 된건가요? 다시 푸시를 한 경우는 전체적으로 반영이 끝난 것인지도 알려주면 좋겠습니다. 
부분만 반영된 경우는 놓치고 넘어갈 수도 있으니까요

wnsxor1993 committed:
Chore: 사각형 색상 및 투명도 변경 기능 수정
- 색상과 투명도 조정에 따른 값 변화 함수 호출 분리
- 우측 속성뷰의 텍스트 표출 함수 간소화

wnsxor1993 committed:
Chore: 제스처 기능 관련 함수 수정
- 뷰를 직접 서칭하지 않고 hit된 뷰가 해당 좌표를 포함하는 모델과 동일한 지 비교하는 형태로 수정

wnsxor1993 commented:
다음에는 푸시할 때 해당 부분도 염두해두겠습니다!
방금 전 푸시가 말씀하신 사항을 최대한 반영하여 수정한 작업물입니다.

godrm commented on DrawingApp/Model/Rectangle/Plane.swift:
```diff
@@ -33,14 +34,10 @@ struct Plane{
-            
-            if withX >= minX, withX <= maxX, withY >= minY, withY <= maxY{
-                findedRectangle = rectangle
+            guard let inRangeRectangle = rectangle.findLocationRange(xPoint: withX, yPoint: withY) else{
+                return
             }
+            findedRectangle = inRangeRectangle
```
forEach로 하면 100개 중에 1,2번째에 있더라도 남은 항목을 다 찾지 않던가요? 이게 좋은 선택일까요?

wnsxor1993 replied:
forEach문에 좋지 않은 집착이 있었던 것 같습니다. 반복에 대한 제어 여부 등도 생각해서 forEach와 for in 중 적합한 요소를 활용할 수 있도록 하겠습니다.

godrm commented on DrawingApp/Model/Attributes/Alpha.swift:
```diff
@@ -22,4 +22,16 @@ enum Alpha: Double, CaseIterable{
         return self.rawValue
     }
+    
+    static func doubleToAlpha(alpha: Double) -> Alpha{
+        var returnAlpha = Alpha.one
+        
+        Alpha.allCases.forEach{ alphaCase in
+            guard alpha >= alphaCase.showValue() else{
+                return
+            }
+            returnAlpha = alphaCase
+        }
+        return returnAlpha
```
doubleToAlpha() 이 함수는 결국 init() 생성자와 같은게 아닐까요?
init(rawValue:) 와 동일한 역할인 것 같은데 어떻게 생각해요?

wnsxor1993 replied:
말씀하신 사항을 보고 init(rawValue:)를 확인하니 생성 값의 옵셔널 여부의 차이가 있을 뿐 동일한 기능을 수행하는 것으로 보입니다. 다만 옵셔널 여부의 차이 또한 해당 메서드 내에서 예외에 대한 값을 고정해서 리턴해주기 때문일 뿐이라 불필요하게 중복되는 기능을 부여한 것 같습니다.

godrm commented on DrawingApp/Model/Rectangle/Rectangle.swift:
```diff
@@ -8,29 +8,17 @@
-        return point
-    }
-    
     func showColor() -> RGBColor{
         return color
     }
```
color값을 가져가야 하는 게 필요하다면 private(set)도 써도 됩니다. 
다만 항상 의식은 해야 합니다. 값을 가져가는 게 전달하기 위해서인가, 아니면 로직을 구현하는 데 여기서 할 수 있는 것을 가져가는게 아닐까 말이죠

godrm commented on DrawingApp/Model/Rectangle/Rectangle.swift:
```diff
@@ -43,28 +31,20 @@ final class Rectangle: CustomStringConvertible{
+            return self
+        } else{
+            return nil
         }
```
여기서 리턴타입으로 self를 넘기는 건 이상하지 않은가요?
좌표값을 기반으로 포함하는 여부만 bool으로 리턴해도 되지 않을까요?

wnsxor1993 replied:
Model에서 자신의 타입의 값이 필요한 로직이 있을 경우에는 해당 모델은 self로 자신의 타입을 넘겨주기 보다는 내부 프로퍼티 타입의 값이나 원시 타입의 값들을 상위 모델에게 넘겨주고, 상위 모델에서 넘겨받은 값들을 가지고 필요한 모델 타입의 값을 찾거나 넘겨주는 형태가 되도록 구현하는 것이 올바른 지향점일까요?

godrm replied:
self 자체를 넘겨주는 일은 없겠죠. 이미 값을 알고 있는 경우가 많을테니까요
속성을 가져가는 경우가 있을텐데 그건 전달하려는 의미가 클꺼구요 (그러면 보통 이런 객체는 로직이 없이 값 타입인 경우가 많아요)
그 외에는 판단해서 결과만 주는 경우가 많을 겁니다. 여기는 좌표가 포함되는지 여부만 리턴해도 될 것 같아요

godrm commented on DrawingApp/VC/MainViewController.swift:
```diff
@@ -0,0 +1,117 @@
+    }
+    
+    private func findSelectedRectangle(point: CGPoint){
+        guard let rectangleInPlane = rectangles.findRectangle(withX: point.x, withY: point.y) else{
+            return
+        }
+        
+        let rectangleView = self.view.hitTest(point, with: nil)
+        
+        if rectangleView?.restorationIdentifier == rectangleInPlane.id{
+            self.selectedRectangleView = rectangleView
+            self.selectedRectangleIndex = rectangles.findRectangleIndex(rectangle: rectangleInPlane)
+        } else{
+            return
+        }
```
뷰를 찾을 때는 rectangle을 찾아서 그것과 매칭하는 것을 뷰 컨트롤러가 갖고 있는 것도 좋습니다
터치 이벤트를 받은 경우는 보통 이미 뷰를 알고 있기는 한데, 서브 뷰들 중에서 따로 찾지 않아도 된다는 의미입니다
이벤트를 받으면 Plane에게 해당하는 Rectangle 모델을 찾으라고 알려주고, 선택되도록 만드는 것만 전달하면 됩니다
여기서 모델도 처리하고 뷰도 처리하고 할 필요는 없습니다. 앞으로는 이 부분을 개선해보세요 (옵저버도 필요할 겁니다)

godrm commented on DrawingApp/View/RandomRectangleView.swift:
```diff
@@ -0,0 +1,28 @@
+        super.init(coder: coder)
+    }
+    
+    func attribute(rectangle: Rectangle){
+        self.backgroundColor = UIColor(red: rectangle.showColor().redValue(), green: rectangle.showColor().greenValue(), blue: rectangle.showColor().blueValue(), alpha: rectangle.showAlpha().showValue())
+        self.restorationIdentifier = rectangle.id
+    }
+    
+    func layout(rectangle: Rectangle){
+        self.frame = CGRect(x: rectangle.point.x, y: rectangle.point.y, width: rectangle.size.width, height: rectangle.size.height)
+    }
```
뷰에 필요한 데이터를 전달할 때도 Rectangle을 전부 전달해야 하는지, 
그걸 생성하는 시점에 전달할 지, 다른 메소드로 전달한다면 어떤 값을 전달할지 역할에 따라 나누는 게 좋습니다
동일한 매개변수를 넘기는 메소드가 두 개가 있는 데 서로 다른 역할을 하는 게 어색합니다
특히 frame 정보는 생성 시점에 처리해도 될꺼 같구요

wnsxor1993 replied:
RightAttributerView와 달리 하위에 생성해야되는 view도 없어 속성과 레이아웃을 따로 나눠줄 필요도 굳이 외부에서 호출해서 속성과 레이아웃을 설정해야 되는 이유도 없는 듯 합니다. 해당 부분은 이니셜라이저를 커스텀해서 생성에 필요한 값만 받아오는 형태로 수정 방향을 잡고자 합니다.

godrm commented on DrawingApp/View/RightAttributerView.swift:
```diff
@@ -0,0 +1,152 @@
+    
+    @objc private func moveBlueSlider(){
+        self.delegate?.moveBlueSlider()
+    }
```
RightAttributerView 선언은 잘 하신 것 같습니다. 
모든 것을 코드로 생성하고 있는데, 코드로 선택한 이유가 있나요?

wnsxor1993 replied:
이전 과제 때는 스토리보드를 이용한 방식을 경험했으니, 이번 과제에서는 코드로 생성하는 방식을 경험하고 학습해야겠다는 생각이 들어, 하위 뷰들을 최대한 코드를 통해 생성하고 있습니다.

godrm commented on DrawingApp/View/RightAttributerView.swift:
```diff
@@ -0,0 +1,152 @@
+        self.colorNameBlue.text = "Blue : \(String(format: "%.0f", blueValue))"
+    }
+    
+    func changeColorSliderValue(){
+        self.colorNameRed.text = "Red : \(String(format: "%.0f", redValue))"
+        self.colorNameGreen.text = "Green : \(String(format: "%.0f", greenValue))"
+        self.colorNameBlue.text = "Blue : \(String(format: "%.0f", blueValue))"
+    }
+    
+    func changeAlphaSliderView(){
+        self.alphaTitle.text = "투명도 : \(String(format: "%.0f", alphaValue.showValue() * 10))"
+    }
```
이 두 메소드는 RightAttributerView에 있는 속성이 바뀐거라면 내부에서 호출해도 될 것 같아보이기도 했습니다
다시보니 출력을 하고 있는 거네요. 그렇다면 ` "Red : \(String(format: "%.0f", redValue))"` 이런 문구도 뷰가 결정하기 보다는 외부에서 결정해서 전달해주는 게 좋습니다
그런 역할이 presentation 역할이 됩니다. 어떤 모델의 속성을 선택하느냐에 따라 값이 바뀌니까 그걸 뷰가 내부에 갖고 있는게 아니라 값이 바뀐 시점에 changeColorSliderValue()를 호출하면서 text에 표시할 값을 전달해주는 게 좋습니다.

godrm reviewed (approved):
개선된 부분이 보이네요. 수고하셨습니다.
다음 피드백을 확인하시고 다음 단계를 하면서 개선해보세요
